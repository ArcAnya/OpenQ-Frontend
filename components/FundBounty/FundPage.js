import React, { useState, useContext } from 'react';
import TokenFundBox from './SearchTokens/TokenFundBox';
import useWeb3 from '../../hooks/useWeb3';
import StoreContext from '../../store/Store/StoreContext';
import { ethers } from 'ethers';
import useConfirmErrorSuccessModals from '../../hooks/useConfirmErrorSuccessModals';
import ConfirmErrorSuccessModalsTrio from '../ConfirmErrorSuccessModals/ConfirmErrorSuccessModalsTrio';
import ButtonLoadingIcon from '../Loading/ButtonLoadingIcon';

const FundPage = ({ bounty, refreshBounty }) => {
	const [volume, setVolume] = useState('');
	const {
		showErrorModal,
		setShowErrorModal,
		showSuccessModal,
		setShowSuccessModal,
		showConfirmationModal,
		setShowConfirmationModal,
	} = useConfirmErrorSuccessModals();
	const [error, setError] = useState('');
	const [isLoading, setIsLoading] = useState(false);
	const [successMessage, setSuccessMessage] = useState('');
	const [transactionHash, setTransactionHash] = useState(null);
	const [confirmationMessage, setConfirmationMessage] = useState('');

	// Context
	const [appState] = useContext(StoreContext);
	const { library, account } = useWeb3();

	// State
	const [token, setToken] = useState(appState.tokens[0]);

	const claimed = bounty.status == 'CLOSED';
	const isLoadingOrIsClosed = isLoading || claimed;
	const disableOrEnable = `${isLoadingOrIsClosed ? 'confirm-btn-disabled cursor-not-allowed' : 'confirm-btn cursor-pointer'}`;
	const fundButtonClasses = `flex flex-row justify-center space-x-5 items-center py-3 text-lg text-white ${disableOrEnable}`;

	// Methods
	async function fundBounty() {
		setIsLoading(true);
		const volumeInWei = volume * 10 ** token.decimals;
		console.log('volumeInWei', volumeInWei);

		if (volumeInWei == 0) {
			setError({ title: 'Zero Volume Sent', message: 'Must send a greater than 0 volume of tokens.' });
			setShowErrorModal(true);
			setIsLoading(false);
			return;
		}

		const bigNumberVolumeInWei = ethers.BigNumber.from(volumeInWei.toString());
		console.log('bigNumberVolumeInWei', bigNumberVolumeInWei);

		let approveSucceeded = false;

		try {
			const callerBalance = await appState.openQClient.balanceOf(library, account, token.address);
			console.log('callerBalance', callerBalance);
			console.log('callerBalance < bigNumberVolumeInWei', callerBalance < bigNumberVolumeInWei);
			console.log('callerBalance.lt(bigNumberVolumeInWei)', callerBalance.lt(bigNumberVolumeInWei));

			if (callerBalance < bigNumberVolumeInWei) {
				const title = 'Funds Too Low';
				const message = 'You do not have sufficient funds for this deposit';
				setError({ message, title });
				setIsLoading(false);
				setShowErrorModal(true);
				return;
			}
		} catch (error) {
			console.log(error);
			const title = 'Error';
			const message = 'A contract call exception occurred';
			setError({ message, title });
			setIsLoading(false);
			setShowErrorModal(true);
			return;
		}

		try {
			if (token.address != ethers.constants.AddressZero) {
				await appState.openQClient.approve(
					library,
					bounty.bountyAddress,
					token.address,
					bigNumberVolumeInWei
				);
			}
			approveSucceeded = true;
		} catch (error) {
			const { message, title } = appState.openQClient.handleError(error, { bounty });
			setError({ message, title });
			setIsLoading(false);
			setShowErrorModal(true);
		}

		if (approveSucceeded) {
			try {
				const fundTxnReceipt = await appState.openQClient.fundBounty(
					library,
					bounty.bountyAddress,
					token.address,
					bigNumberVolumeInWei
				);
				setTransactionHash(fundTxnReceipt.transactionHash);
				setSuccessMessage(
					`Successfully funded issue ${bounty.url} with ${volume} ${token.symbol}!`
				);
				setShowSuccessModal(true);
				refreshBounty();
				setIsLoading(false);
			} catch (error) {
				const { message, title } = appState.openQClient.handleError(error, { bounty });
				setError({ message, title });
				setIsLoading(false);
				setShowErrorModal(true);
			}
		}
	}

	function onCurrencySelect(token) {
		setToken(token);
		setConfirmationMessage(
			`You are about to fund this bounty at address ${bounty.bountyAddress.substring(
				0,
				12
			)}...${bounty.bountyAddress.substring(32)} with ${volume} ${token.name
			}. Is this correct?`
		);
	}

	function onVolumeChange(volume) {
		setVolume(volume);
		setConfirmationMessage(
			`You are about to fund this bounty at address ${bounty.bountyAddress.substring(
				0,
				12
			)}...${bounty.bountyAddress.substring(32)} with ${volume} ${token.name
			}. Is this correct?`
		);
	}

	//Close Modal on outside click
	/* useEffect(() => {
		let handler = (event) => {
			if (!menuRef.current.contains(event.target)) {
				if (!isLoading) {
					if (!notifyMenuRef.current.contains(event.target)) {
						setIsLoading(false);
						updateModal();
					}
				} else {
					updateModal();
				}
			}
		};
		window.addEventListener("mousedown", handler);

		return () => {
			window.removeEventListener("mousedown", handler);
		};
	}); */

	// Render
	if (claimed) {
		return (
			<div className="pt-16">
				<div className="flex flex-col space-y-5">
					<div className="bg-purple-600 col-span-3 bg-opacity-20 border border-purple-700 rounded-lg text-white p-4">
						Bounty Is Already Closed
					</div>
				</div>
			</div>
		);
	} else {
		return (
			<div className="flex flex-1 font-mont justify-center items-center">
				<div className="flex flex-col space-y-5 w-5/6">
					<div className="flex text-3xl font-semibold text-white justify-center pt-16">
						Fund Bounty
					</div>
					<div className="bg-purple-600 col-span-3 bg-opacity-20 border border-purple-700 rounded-lg text-white p-4">
						Deposited ERC-20 Tokens can be withdrawn again after 30 days
					</div>

					<TokenFundBox
						onCurrencySelect={onCurrencySelect}
						onVolumeChange={onVolumeChange}
						token={token}
						volume={volume}
					/>

					<div>
						<button
							className={fundButtonClasses}
							disabled={isLoading || claimed}
							type="button"
							onClick={() => setShowConfirmationModal(true)}
						>
							<div>{!isLoading ? 'Fund' : 'Approving'}</div>
							<div>{isLoading && <ButtonLoadingIcon />}</div>
						</button>
					</div>
					{/*  <div className="flex items-center justify-end p-6 border-solid border-blueGray-200 rounded-b">
								<button
									className="text-red-500 background-transparent font-bold uppercase px-6 py-2 text-sm outline-none focus:outline-none mr-1 mb-1 ease-linear transition-all duration-150"
									type="button"
									onClick={() => updateModal()}
								>
									Close
								</button>
							</div> */}
				</div>

				<ConfirmErrorSuccessModalsTrio
					setShowErrorModal={setShowErrorModal}
					showErrorModal={showErrorModal}
					error={error}
					setShowConfirmationModal={setShowConfirmationModal}
					showConfirmationModal={showConfirmationModal}
					confirmationTitle={'Confirm Deposit'}
					confirmationMessage={confirmationMessage}
					positiveOption={'Approve'}
					confirmMethod={fundBounty}
					showSuccessModal={showSuccessModal}
					setShowSuccessModal={setShowSuccessModal}
					successMessage={successMessage}
					transactionHash={transactionHash}
				/>
			</div>
		);
	}
};

export default FundPage;
