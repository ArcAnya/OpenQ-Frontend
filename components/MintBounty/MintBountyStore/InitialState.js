const InitialState = {
	bounty: null,
	bountyAddress: '',
	bountyExists: false,
	walletConnected: false,
	enableMint: false,
	error: null,
	errorMessage: '',
	isBountyMinted: false,
	isLoading: false,
	isValidUrl: false,
	issueClosed: false,
	issueData: '',
	issueFound: false,
	issueId: '',
	issueNumber: '',
	issueUrl: '',
	orgId: '',
	repoName: '',
	transactionPending: false
};

export default InitialState;