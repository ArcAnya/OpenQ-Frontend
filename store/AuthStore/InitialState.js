const InitialState = {
    isAuthenticated: false,
    user: null,
    token: null,
};

export default InitialState;