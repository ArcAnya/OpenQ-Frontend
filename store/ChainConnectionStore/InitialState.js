const InitialState = {
    hasMetamask: false,
    activeAccount: null,
    showWalletConnect: true,
    fullyConnected: false,
    isOnCorrectNetwork: false,
};

export default InitialState;