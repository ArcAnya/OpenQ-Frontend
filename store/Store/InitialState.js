import Utils from '../../services/utils/Utils';

import OpenQClient from '../../services/ethers/OpenQClient';
import OpenQSubgraphClient from '../../services/subgraph/OpenQSubgraphClient';
import GithubRepository from '../../services/github/GithubRepository';
import TokenClient from '../../services/coins/TokenClient';
import GithubBot from '../../services/github-bot/GithubBot';
import Logger from '../../services/logger/Logger';

import MockGithubRepository from '../../services/github/MockGithubRepository';
import MockOpenQClient from '../../services/ethers/MockOpenQClient';
import MockOpenQSubgraphClient from '../../services/subgraph/MockOpenQSubgraphClient';
import MockTokenClient from '../../services/coins/MockTokenClient';
import MockGithubBot from '../../services/github-bot/MockGithubBot';

// Token Metadata
// Array of all supported tokens
import polygonMainnetTokens from '../../constants/polygon-mainnet-tokens.json';
import mumbaiTokens from '../../constants/polygon-mumbai-tokens.json';
import localTokens from '../../constants/local-tokens.json';

// Mapping of tokens with token metadata for token address lookup
import polygonMainnetTokenMetadata from '../../constants/polygon-mainnet.json';
import mumbaiTokenMetadata from '../../constants/polygon-mumbai.json';
import localTokenMetadata from '../../constants/local.json';

let InitialState = {};
switch (process.env.NEXT_PUBLIC_DEPLOY_ENV) {
	case 'local':
		InitialState = {
			tokenMetadata: localTokenMetadata,
			tokens: localTokens,
			openQClient: new MockOpenQClient(),
			githubRepository: new MockGithubRepository(),
			openQSubgraphClient: new MockOpenQSubgraphClient(),
			tokenClient: new MockTokenClient(),
			githubBot: new MockGithubBot(),
			logger: new Logger(true),
			utils: new Utils(),
		};
		break;
	case 'docker':
		InitialState = {
			tokenMetadata: localTokenMetadata,
			tokens: localTokens,
			openQClient: new OpenQClient(),
			githubRepository: new GithubRepository(),
			openQSubgraphClient: new OpenQSubgraphClient(),
			tokenClient: new TokenClient(),
			githubBot: new MockGithubBot(),
			logger: new Logger(true),
			utils: new Utils(),
		};
		break;
	case 'development':
		InitialState = {
			tokenMetadata: mumbaiTokenMetadata,
			tokens: mumbaiTokens,
			openQClient: new OpenQClient(),
			githubRepository: new GithubRepository(),
			openQSubgraphClient: new OpenQSubgraphClient(),
			tokenClient: new TokenClient(),
			githubBot: new MockGithubBot(),
			logger: new Logger(true),
			utils: new Utils(),
		};
		break;
	case 'staging':
		InitialState = {
			tokenMetadata: polygonMainnetTokenMetadata,
			tokens: polygonMainnetTokens,
			openQClient: new OpenQClient(),
			githubRepository: new GithubRepository(),
			openQSubgraphClient: new OpenQSubgraphClient(),
			tokenClient: new TokenClient(),
			githubBot: new GithubBot(),
			logger: new Logger(true),
			utils: new Utils(),
		};
		break;
	case 'production':
		InitialState = {
			tokenMetadata: polygonMainnetTokenMetadata,
			tokens: polygonMainnetTokens,
			openQClient: new OpenQClient(),
			githubRepository: new GithubRepository(),
			openQSubgraphClient: new OpenQSubgraphClient(),
			tokenClient: new TokenClient(),
			githubBot: new GithubBot(),
			logger: new Logger(false),
			utils: new Utils(),
		};
		break;
	default:
		throw Error('ENVIRONMENT NOT CONFIGURED CORRECTLY. Set an environment with DEPLOY_ENV');
}

export default InitialState;