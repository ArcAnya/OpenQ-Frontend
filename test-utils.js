// Third party
import React from 'react';
// Cusotm
import { render } from '@testing-library/react';
import StoreProvider from './store/Store/StoreProvider';
import AuthProvider from './store/AuthStore/AuthProvider';

// Add in any providers here if necessary:
// test-utils.js
// (ReduxProvider, ThemeProvider, etc)
const Providers = ({ children }) => {
  return (
    <AuthProvider>
      <StoreProvider>{children}</StoreProvider>
    </AuthProvider>
  );
};

const customRender = (ui, options = {}) => render(ui, { wrapper: Providers, ...options });

// re-export everything
export * from '@testing-library/react';

// override render method
export { customRender as render };
