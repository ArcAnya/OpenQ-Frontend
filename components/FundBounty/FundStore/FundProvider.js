import React, { useReducer } from 'react';
import FundReducer from './FundReducer';
import FundContext from '../FundStore/FundContext';
import InitialState from './InitialFundState.js';

const FundProvider = ({ children, bounty, refreshBounty }) => {
  const [state, dispatch] = useReducer(FundReducer, { bounty, refreshBounty, ...InitialState });

  return <FundContext.Provider value={[state, dispatch]}>{children}</FundContext.Provider>;
};

export default FundProvider;
