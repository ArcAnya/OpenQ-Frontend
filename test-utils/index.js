// Third party
import React from 'react';
// Cusotm
import { render, renderHook } from '@testing-library/react';

import StoreProvider from '../store/Store/TestStoreProvider';
import AuthProvider from '../store/AuthStore/TestAuthProvider';
import FundProvider from '../components/FundBounty/TestFundProvider';

// Add in any providers here if necessary:
// test-utils.js
// (ReduxProvider, ThemeProvider, etc)
const customRender = (ui, options = {}, storeProps, authProps) => {
  const Providers = ({ children }) => {
    return (
      <AuthProvider StoreProps={authProps}>
        <StoreProvider StoreProps={storeProps}>{children}</StoreProvider>
      </AuthProvider>
    );
  };

  render(ui, { wrapper: Providers, ...options });
};
const customHookRender = (hook, options = {}, storeProps, authProps, fundProps) => {
  const Providers = ({ children }) => {
    return (
      <AuthProvider StoreProps={authProps}>
        <StoreProvider StoreProps={storeProps}>
          <FundProvider StoreProps={fundProps}> {children}</FundProvider>
        </StoreProvider>
      </AuthProvider>
    );
  };

  return renderHook(hook, { wrapper: Providers, ...options });
};

// re-export everything
export * from '@testing-library/react';

// override render method
export { customRender as render, customHookRender as hookRender };
