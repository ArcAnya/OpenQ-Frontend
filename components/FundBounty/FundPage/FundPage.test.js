/**
 * @jest-environment jsdom
 */
import React from 'react';
import { render, screen } from '../../../test-utils';
import FundPage from '.';
import userEvent from '@testing-library/user-event';
import InitialState from '../../../store/Store/InitialState';
import FundContext from '../../FundBounty/FundStore/FundContext';
import FundProvider from '../../FundBounty/FundStore/FundProvider';
import Constants from '../../../test-utils/constant';
import InitialFundState from '../../FundBounty/FundStore/InitialFundState';
import { CONFIRM } from '../../FundBounty/FundStore/ApproveFundState';

describe('FundPage', () => {
  const bounty = Constants.bounty;
  const splitBounty = Constants.bounty1;

  const refreshBounty = () => {
    return null;
  };
  beforeEach(() => {
    const observe = jest.fn();
    const disconnect = jest.fn();

    window.IntersectionObserver = jest.fn(() => ({
      observe,
      disconnect,
    }));

    InitialState.openQClient.reset();
    InitialState.shouldSleep = 200;
  });

  describe('Split Price Bounties', () => {
    it('should lock token choice to deposit token when there is 1+ deposit', async () => {
      // ARRANGE
      const otherTokenDepositSplitBounty = { ...splitBounty, deposits: [Constants.deposit3] };
      const user = userEvent.setup();
      const { asFragment } = render(
        <FundProvider bounty={otherTokenDepositSplitBounty} refreshBounty={refreshBounty}>
          <FundPage bounty={otherTokenDepositSplitBounty} />
        </FundProvider>
      ); // ACT / ASSERT
      expect(asFragment()).toMatchInlineSnapshot(`
        <DocumentFragment>
          <div
            class="flex-1 pt-4 pb-8 w-full max-w-[1200px] justify-center"
          >
            <div
              class="flex flex-col w-full space-y-5 pb-8 items-center md:border rounded-sm border-gray-700"
            >
              <div
                class="flex text-3xl w-full text-primary justify-center px-16 py-4 md:bg-[#161b22] md:border-b border-gray-700 rounded-t-sm"
              >
                Escrow Funds in Split Price Contract
              </div>
              <div
                class="flex flex-col space-y-5 w-5/6 pt-2"
              >
                <div
                  class="flex w-full gap-4"
                >
                  <div
                    class="flex space-x-4 w-full undefined"
                  >
                    <div
                      class="flex w-full flex-row justify-between items-center px-4 input-field-big"
                    >
                      <div
                        class=" bg-dark-mode"
                      >
                        <input
                          aria-label="amount"
                          autocomplete="off"
                          class="font-semibold number outline-none bg-input-bg text-primary w-full"
                          id="amount"
                          placeholder="0.0"
                          value=""
                        />
                      </div>
                    </div>
                    <div
                      class="flex"
                    >
                      <button
                        aria-label="select token"
                        class="flex flex-row items-center p-0.5 px-2 cursor-default"
                        disabled=""
                      >
                         
                        <div
                          class="flex h-4 w-4 items-center justify-center"
                        >
                          <img
                            alt="selected token"
                            class="rounded-full"
                            data-nimg="1"
                            decoding="async"
                            height="40"
                            loading="lazy"
                            src="https://wallet-asset.matic.network/img/tokens/link.svg"
                            srcset="https://wallet-asset.matic.network/img/tokens/link.svg 1x, https://wallet-asset.matic.network/img/tokens/link.svg 2x"
                            style="color: transparent;"
                            width="40"
                          />
                        </div>
                        <div
                          class="flex pl-2 pr-1 text-primary"
                        >
                          LINK
                        </div>
                      </button>
                    </div>
                  </div>
                </div>
                <div
                  class="flex gap-4"
                >
                  <div
                    class="w-full"
                  >
                    <div
                      class="flex w-full input-field-big mb-4"
                    >
                      <div
                        class=" flex items-center gap-3 w-full text-primary md:whitespace-nowrap"
                      >
                        <div
                          class="group "
                        >
                          <div
                            class="cursor-help rounded-full border border-[#c9d1d9] aspect-square leading-4 h-4 box-content text-center font-bold text-primary"
                          >
                            ?
                          </div>
                          <div
                            class="justify-center w-full relative hidden z-50 group-hover:block  -top-1 "
                          >
                            <div
                              class="flex flex-col items-center"
                            >
                              <div
                                class="flex mt-0.5 md:mt-1 tooltip-triangle absolute undefined"
                              />
                              <div
                                class="flex tooltip absolute md:-left-12"
                              >
                                <div
                                  class="whitespace-normal md:w-96 sm:w-60 w-40  "
                                >
                                  This is the number of days that your deposit will be in escrow. After this many days, your deposit will be fully refundable if the contract has still not been claimed.
                                </div>
                              </div>
                            </div>
                          </div>
                        </div>
                        <span>
                          Deposit Locked Period
                        </span>
                      </div>
                      <div
                        class="flex px-4 font-bold bg-input-bg"
                      >
                        <input
                          autocomplete="off"
                          class="text-primary text-right number outline-none w-full flex-1 bg-input-bg"
                          id="deposit-period"
                          value="30"
                        />
                      </div>
                    </div>
                    <div
                      class="flex items-center"
                    />
                    <div
                      class="group w-min"
                    >
                      <button
                        class="text-center px-8 w-min items-center py-0.5  btn-default w-full cursor-not-allowed py-1.5"
                        disabled=""
                        type="button"
                      >
                        <div
                          class="text-center whitespace-nowrap w-full"
                        >
                          Fund
                        </div>
                      </button>
                      <div
                        class="justify-center w-full relative hidden z-50 group-hover:block  -top-1  "
                      >
                        <div
                          class="flex flex-col items-center"
                        >
                          <div
                            class="flex mt-0.5 md:mt-1 tooltip-triangle absolute undefined"
                          />
                          <div
                            class="flex tooltip absolute left-0"
                          >
                            <div
                              class="sm:w-40 md:w-60 whitespace-normal"
                            >
                              Please indicate the volume you'd like to fund with. Must be between 0.0000001 and 1,000,000.
                            </div>
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </DocumentFragment>
      `);

      const input = screen.getByLabelText('amount');
      await user.type(input, '0.30sdf');
      expect(screen.getAllByText(/link/i)[0]).toBeInTheDocument;
      const selectToken = screen.getByRole('button', { name: 'select token' });
      expect(selectToken).toBeDisabled();
      expect(selectToken).toMatchSnapshot('Snapshot: button style changes to not appear as button');
      const button = screen.getByRole('button', { name: /Fund/i });
      await user.click(button);
    });
  });

  describe('Fixed Bounties', () => {
    it('should render the heading', async () => {
      // ARRANGE
      const user = userEvent.setup();
      render(
        <FundProvider bounty={bounty} refreshBounty={refreshBounty}>
          <FundPage bounty={bounty} />
        </FundProvider>
      );

      // ACT
      const heading = screen.getByText('Fund');
      expect(heading).toBeInTheDocument();
      const input = screen.getByLabelText('amount');
      await user.type(input, '200');
      const button = screen.getByRole('button', { name: /Fund/i });
      await user.click(button);
      const confirmBtn = await screen.findAllByRole('button', { name: /Fund/i });
      await user.click(confirmBtn[1]);
      const modalContent = await screen.findByText(/Too Low/i);

      // ASSERT
      expect(modalContent).toBeInTheDocument();
      await user.click(screen.getByRole('button', { name: 'Close' }));
      expect(modalContent).not.toBeInTheDocument();
      const nullish = [...screen.queryAllByRole(/null/), ...screen.queryAllByRole(/undefined/)];
      expect(nullish).toHaveLength(0);
    });

    it('should render list items', async () => {
      const user = userEvent.setup();
      // ARRANGE
      render(
        <FundProvider bounty={bounty} refreshBounty={refreshBounty}>
          <FundPage bounty={bounty} />
        </FundProvider>
      );
      const input = screen.getByLabelText('amount');
      await user.type(input, '200');

      // ACT
      const token = screen.getByText('MATIC');
      const button = screen.getByRole('button', { name: /Fund/i });
      await user.click(button);
      const confirmBtn = await screen.findAllByRole('button', { name: /Fund/i });
      expect(confirmBtn[0]).toBeInTheDocument();
      await user.click(confirmBtn[1]);
      const modalContent = await screen.findByText(/Too Low/i);

      // ASSERT
      expect(modalContent).toBeInTheDocument();

      // ASSERT
      expect(token).toBeInTheDocument();
      const nullish = [...screen.queryAllByRole(/null/), ...screen.queryAllByRole(/undefined/)];
      expect(nullish).toHaveLength(0);
    });

    it('should let user submit and handle too low amount of token', async () => {
      // ARRANGE
      const user = userEvent.setup();
      render(
        <FundProvider bounty={bounty} refreshBounty={refreshBounty}>
          <FundPage bounty={bounty} />
        </FundProvider>
      );
      // ACT
      const input = screen.getByLabelText('amount');
      await user.type(input, '200');
      const button = screen.getByRole('button', { name: /Fund/i });
      await user.click(button);
      const confirmBtn = await screen.findAllByRole('button', { name: /Fund/i });
      await user.click(confirmBtn[1]);
      const modalContent = await screen.findByText(/Too Low/i);

      // ASSERT
      expect(modalContent).toBeInTheDocument();
      await user.click(screen.getByRole('button', { name: 'Close' }));
      expect(modalContent).not.toBeInTheDocument();
      const nullish = [...screen.queryAllByRole(/null/), ...screen.queryAllByRole(/undefined/)];
      expect(nullish).toHaveLength(0);
    });

    it('should let user submit and handle owned amount of Matic', async () => {
      // ARRANGE
      const user = userEvent.setup();
      const state = {
        ...InitialFundState,
        bounty,
        refreshBounty,
        token: Constants.tokenMetadata,
        volume: '3',
        approveTransferState: CONFIRM,
      };
      const dispatch = jest.fn();
      render(
        <FundContext.Provider value={[state, dispatch]}>
          <FundPage bounty={bounty} />
        </FundContext.Provider>
      );
      // ACT
      const value = await screen.findByText(/3 Dai/i);

      // ASSERT
      expect(value).toBeInTheDocument();
      const confirmBtn = await screen.findAllByRole('button', { name: /Fund/i });
      await user.click(confirmBtn[1]);
      expect(dispatch).toHaveBeenCalledWith({ type: 'SET_APPROVE_TRANSFER_STATE', payload: 'APPROVE' });
      const nullish = [...screen.queryAllByRole(/null/), ...screen.queryAllByRole(/undefined/)];
      expect(nullish).toHaveLength(0);
    });

    it('should let user submit and handle owned amount of Link', async () => {
      // ARRANGE
      const user = userEvent.setup();
      render(
        <FundProvider bounty={bounty} refreshBounty={refreshBounty}>
          <FundPage bounty={bounty} />
        </FundProvider>
      ); // ACT
      const input = screen.getByLabelText('amount');
      await user.type(input, '0.30sdf');
      await user.click(screen.getByText(/Matic/i));
      await user.click(screen.getAllByText(/link/i)[0]);
      const button = screen.getByRole('button', { name: /Fund/i });
      await user.click(button);
      const value = await screen.findByText(/.3 link/i);

      // ASSERT
      expect(value);
      const confirmBtn = await screen.findByRole('button', { name: /Approv/i });
      await user.click(confirmBtn);
      const funding = await screen.findByText('Funding...');
      expect(funding).toBeInTheDocument();
      const modalContent = await screen.findByText(/Transfer Complete/i, undefined, {
        timeout: 8000,
      });
      const tweet = screen.getByText('Tweet about it');
      expect(tweet).toBeInTheDocument();
      await user.click(screen.getByTestId('cross'));
      expect(modalContent).not.toBeInTheDocument();
      const nullish = [...screen.queryAllByRole(/null/), ...screen.queryAllByRole(/undefined/)];
      expect(nullish).toHaveLength(0);
    });

    it('should go straight to fund when DERC20 approved previously', async () => {
      // ARRANGE
      const user = userEvent.setup();
      render(
        <FundProvider bounty={bounty} refreshBounty={refreshBounty}>
          <FundPage bounty={bounty} />
        </FundProvider>
      );
      // ACT
      const input = screen.getByLabelText('amount');
      await user.type(input, '4.8');
      await user.click(screen.getByText(/Matic/i));
      await user.click(screen.getAllByText(/derc20/i)[0]);
      const button = screen.getByRole('button', { name: /Fund/i });
      await user.click(button);
      const value = await screen.findByText(/4.8 derc20/i);

      // ASSERT
      expect(value).toBeInTheDocument();
      const confirmBtn = await screen.findAllByRole('button', { name: /Fund/i });
      await user.click(confirmBtn[1]);
      const modalContent = await screen.findByText(/Transfer Complete/i);
      const tweet = screen.getByText('Tweet about it');
      expect(tweet).toBeInTheDocument();
      await user.click(screen.getByTestId('cross'));
      expect(modalContent).not.toBeInTheDocument();
      const nullish = [...screen.queryAllByRole(/null/), ...screen.queryAllByRole(/undefined/)];
      expect(nullish).toHaveLength(0);
    });

    it('should handle approval errors', async () => {
      // ARRANGE
      InitialState.openQClient.shouldError = true;
      const user = userEvent.setup();
      render(
        <FundProvider bounty={bounty} refreshBounty={refreshBounty}>
          <FundPage bounty={bounty} />
        </FundProvider>
      );
      // ACT
      const input = screen.getByLabelText('amount');
      await user.type(input, '0.30sdf');
      await user.click(await screen.findByText(/MATIC/i));
      await user.click(screen.getAllByText(/Chainlink/i)[0]);
      const button = await screen.findByRole('button', { name: /Fund/i });
      await user.click(button);

      // ASSERT
      expect(await screen.findByText(/0.3 LINK/));
      await user.click(await screen.findByRole('button', { name: /Approv/i }));
      const modalContent = await screen.findByText(/try again./i);
      expect(modalContent).toBeInTheDocument();

      await user.click(await screen.findByRole('button', { name: 'Close' }));
      expect(modalContent).not.toBeInTheDocument();
      const nullish = [...screen.queryAllByRole(/null/), ...screen.queryAllByRole(/undefined/)];
      expect(nullish).toHaveLength(0);
    });

    it('should prevent user from submitting over 10000000', async () => {
      // ARRANGE
      const user = userEvent.setup();
      render(
        <FundProvider bounty={bounty} refreshBounty={refreshBounty}>
          <FundPage bounty={bounty} />
        </FundProvider>
      );
      // ACT
      const input = screen.getByLabelText('amount');
      await user.type(input, '10000000');
      const button = await screen.findByRole('button', { name: /Fund/i });

      // ASSERT
      expect(button).toBeDisabled();
      await user.hover(button);
      const tooltip = await screen.findByText(/Must be between/);
      expect(tooltip).toBeInTheDocument();
      const nullish = [...screen.queryAllByRole(/null/), ...screen.queryAllByRole(/undefined/)];
      expect(nullish).toHaveLength(0);
    });

    it('should prevent user from submitting under 0.00000001', async () => {
      // ARRANGE
      const user = userEvent.setup();
      render(
        <FundProvider bounty={bounty} refreshBounty={refreshBounty}>
          <FundPage bounty={bounty} />
        </FundProvider>
      );
      // ACT
      const input = screen.getByLabelText('amount');
      await user.type(input, '0.00000001');
      const button = await screen.findByRole('button', { name: /Fund/i });

      // ASSERT
      expect(button).toBeDisabled();
      await user.hover(button);
      const tooltip = await screen.findByText(/Must be between/);
      expect(tooltip).toBeInTheDocument();
      const nullish = [...screen.queryAllByRole(/null/), ...screen.queryAllByRole(/undefined/)];
      expect(nullish).toHaveLength(0);
    });

    it('should show tooltip', async () => {
      // ARRANGE
      const user = userEvent.setup();
      render(
        <FundProvider bounty={bounty} refreshBounty={refreshBounty}>
          <FundPage bounty={bounty} />
        </FundProvider>
      ); // ACT / ASSERT
      const button = screen.getByRole('button', { name: /Fund/i });
      expect(button).toBeDisabled();
      await user.hover(button);
      const tooltip = await screen.findByText(/indicate the volume you'd like to fund with./i);
      expect(tooltip).toBeInTheDocument();
      const nullish = [...screen.queryAllByRole(/null/), ...screen.queryAllByRole(/undefined/)];
      expect(nullish).toHaveLength(0);
    });
  });
});
