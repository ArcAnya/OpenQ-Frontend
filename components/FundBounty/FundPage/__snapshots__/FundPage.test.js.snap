// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FundPage Split Price Bounties should lock token choice to deposit token when there is 1+ deposit: Snapshot: button style changes to not appear as button 1`] = `
<button
  aria-label="select token"
  class="flex flex-row items-center p-0.5 px-2 cursor-default"
  disabled=""
>
   
  <div
    class="flex h-4 w-4 items-center justify-center"
  >
    <img
      alt="selected token"
      class="rounded-full"
      data-nimg="1"
      decoding="async"
      height="40"
      loading="lazy"
      src="https://wallet-asset.matic.network/img/tokens/link.svg"
      srcset="https://wallet-asset.matic.network/img/tokens/link.svg 1x, https://wallet-asset.matic.network/img/tokens/link.svg 2x"
      style="color: transparent;"
      width="40"
    />
  </div>
  <div
    class="flex pl-2 pr-1 text-primary"
  >
    LINK
  </div>
</button>
`;
